server/fleet/cron_schedules.go

```diff
@@ -0,0 +1,141 @@
+package fleet
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+)
+
+type CronSchedulesService interface {
+	// TriggerCronSchedule attempts to trigger an ad-hoc run of the named cron schedule.
+	TriggerCronSchedule(name string) error
+	// // GetCronScheduleNames returns a list of the names of all cron schedules registered with the service.
+	// GetCronScheduleNames() []string
+}
+
+type CronSchedule interface {
+	Trigger() (bool, *CronStats, error)
+	Name() string
+}
+
+type CronSchedules struct {
+	Schedules map[string]CronSchedule
+}
+
+// AuthzType implements authz.AuthzTyper.
+func (cs *CronSchedules) AuthzType() string {
+	return "cron_schedules"
+}
+
+// AddCronSchedules registers a new cron schedule with the service.
+func (cs *CronSchedules) AddCronSchedule(sched CronSchedule, err error) error {
+	if err != nil {
+		return err
+	}
+	cs.Schedules[sched.Name()] = sched
+	return nil
+}
+
+// TriggerCronSchedule attempts to trigger an ad-hoc run of the named cron schedule.
+func (cs *CronSchedules) TriggerCronSchedule(name string) error {
+	sched, ok := cs.Schedules[name]
+	if !ok {
+		return triggerNotFoundError{}
+	}
+	ok, stats, err := sched.Trigger()
+	switch {
+	case err != nil:
+		return err
+	case !ok:
+		if stats == nil || string(stats.Status) == "" {
+			return triggerConflictError{name: name}
+		}
+		return triggerConflictError{name: name, stats: stats}
```

Comment:
Why do we have `err` and `ok`? If things are not ok, we might still return `err==nil`? Maybe instead of ok, you meant something like `ran`?

Reply:
For this method, `!ok` means that nothing was sent on the trigger channel, which could happen in the following cases:
- error getting the current status from the db
- current status is pending
- trigger channel is full

This is the method that gets called by the endpoint handler so we'll need to be returning before we know if we completed the run or not (similar to how we handle the refetch endpoint). I suppose we could call it "signaled" or something like that rather than "ok" if you think that would be more clear.

Reply:
I thought about handling it that way, but it felt like a situation where the method closest to the endpoint should be the place to differentiate how the cases should be presented (i.e. whether to promote `!ok && err == nil` to a user-facing HTTP error). Here's the underlying `schedule.Trigger()` [method](https://github.com/fleetdm/fleet/pull/8747/files#diff-4d9802305cc9d3122103715e85d3f17967232c5ad481d6a87598a52e22e05f74R340) if you'd like to take a look at that and let me know your thoughts on the API.

Reply:
So in this case:

```
default:
		level.Debug(s.logger).Log("msg", "trigger channel not available")
		return false, nil, nil
```

It means "somebody already triggered and it happened that this instance was the one that the load balancer sent that too, and the prev trigger is still waiting on to be executed, so go away".

I think I would just do something like:

```

func (s *Schedule) Trigger() (bool, *fleet.CronStats, error) {
	stats, err := s.getLatestStats()
	if err != nil {
		return nil, err
	}
	if stats.Status == fleet.CronStatsStatusPending {
		return &stats, nil
	}

	go func() {
		select {
		case s.trigger <- struct{}{}:
		case <-<wait on a ticker here for a few seconds>:
                       // maybe log something here
		}
	}()

	return nil, nil
}
```

Because the buffer idea is good, but only if you are running 1 fleet instance, otherwise I think it's the same as not having the buffer at all.

Reply:
> ```
> case <-<wait on a ticker here for a few seconds>:
>                        // maybe log something here
> 		}
> ```

So if this times out (meaning we weren't able to send the trigger on the channel), we return `nil, nil`? To me, that implies to the user that sending the signal and not sending the signal because the channel is blocked (or in theory if the channel doesn't exist) are all the same thing. Is that what you mean imply?

Reply:
You try to send the signal. If you can't then the schedule is already about to be triggered. Right?

Reply:
Most likely. I can't readily imagine exceptions (other than a non-existent channel, which shouldn't happen because we make it by default in `schedule.New`). So if we're not worried about weird edge cases (beyond simply logging them), I'm not sure why we'd want the select to wait for a timeout rather than just proceed immediately through a default case and exiting with `nil, nil`.

Reply:
That makes sense.